tests/quality_pragmaitc/test_refinement.py

```diff
@@ -24,63 +32,430 @@ def test_shouldBeApplicable():
+    task2 = Task()
+
+    current = Context("C1")
+    fullContext = []
+    fullContext.append(current)
+
+    qc = QualityConstraint(current, CommonMetrics.SECONDS,
+                           15, Comparison.LESS_OR_EQUAL_TO)
+
+    task1.addApplicableContext(current)
+    task1.setProvidedQuality(current, CommonMetrics.SECONDS, 16)
+
+    task2.addApplicableContext(current)
+    task2.setProvidedQuality(current, CommonMetrics.SECONDS, 11)
+
+    goal.addDependency(task1)
+    goal.addDependency(task2)
+
+    goal.setIdentifier("Root")
+    goal.addApplicableContext(current)
+
+    interp = Interpretation()
+    interp.addQualityConstraint(qc)
+
+    plan = goal.isAchievable(fullContext, interp)
+    assert plan.getTasks().contains(task2)
+    assert plan.getTasks().contains(task1)
+
+
+def test_aGoalOrDecomposedWithTwoTasksMayNotBeAchievable():
+    goal = Goal(Decomposition.OR)
+
+    task1 = Task()
+    task2 = Task()
+    current = Context("C1")
+    fullContext = []
+    fullContext.append(current)
+
+    qc = QualityConstraint(current, CommonMetrics.SECONDS,
+                           15, Comparison.LESS_OR_EQUAL_TO)
+
+    task1.addApplicableContext(current)
+    task1.setProvidedQuality(current, CommonMetrics.SECONDS, 16)
+
+    task2.addApplicableContext(current)
+    task2.setProvidedQuality(current, CommonMetrics.SECONDS, 17)
+
+    goal.addDependency(task1)
+    goal.addDependency(task2)
+
+    goal.setIdentifier("Root")
+    goal.addApplicableContext(current)
+
+    interp = Interpretation()
+    interp.addQualityConstraint(qc)
+
+    plan = goal.isAchievable(fullContext, interp)
+    assert goal.decomposition
+    assert plan is None
+
+
+def test_ApplicableDeps():
+    goal = Pragmatic()
+
+    task = Task()
+    context = Context("C1")
+    wrongContext = Context("C2")
+    current = []
+
+    qc = QualityConstraint(context, CommonMetrics.SECONDS,
+                           15, Comparison.LESS_OR_EQUAL_TO)
 
-    def test_aTaskShouldBeAchievable():
-        commonMetrics = CommonMetrics()
-        task = Task()
+    task.addApplicableContext(context)
+    task.setProvidedQuality(context, CommonMetrics.SECONDS, 13)
 
-        currentContext = Context("C1")
-        fullContext = []
-        fullContext.append(currentContext)
+    goal.addDependency(task)
+    goal.setIdentifier("Root")
+    goal.addApplicableContext(context)
+    goal.getInterpretation().addQualityConstraint(qc)
 
-        qc = QualityConstraint(currentContext, commonMetrics.SECONDS, 15, 'LESS_OR_EQUAL_TO')
+    interp = Interpretation()
+    interp.addQualityConstraint(qc)
 
-        task.addApplicableContext(currentContext)
-        task.setProvidedQuality(currentContext, commonMetrics.SECONDS, 12)
+    current.append(wrongContext)
+    assert goal.isAchievable(current, interp) is None
 
-        interp = Interpretation()
-        interp.addQualityConstraint(qc)
+    current.append(context)
+    assert len(goal.isAchievable(current, interp).getTasks()) == 1
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/gh/lesunb/p2psimpy/pullRequest?prid=4861978)